CONTRIBUTING.md

```diff
@@ -0,0 +1,75 @@
+   - Applies milestone for next release if the PM thinks there is enough time for the PR to be merged and sufficiently tested on `master` before code complete. Otherwise moved to a future release, letting submitter know that PR may have a delay in review due to the release cycle.
+ - Follows up with contributor if the CLA has not been signed. If no response within 7 days, PM closes the issue.
+ - PM verifies there is a corresponding JIRA ticket or GitHub issue.
+
+Next, the PM tests changes on a spinmint test server:
+ - PM tests and verifies pull requests via the `Setup Test Server` label. Initial review is completed within 2 business days.
+ - If changes are required, PM submits review as "Changes Requested", with a comment on the areas that require updates. Comment explains why changes are needed linking back to design principles.
+   - PM applies `Awaiting Submitter Action` label to more easily query the PR queue.
+   - Once changes are made, PM regenerates test server and repeats testing.
+ - If bugs are found that are also on `master`, a new bug report is submitted in JIRA and linked to the PR. Bugs that are also found on `master` will typically not block merging of PRs.
+ - If PR is approved, PM submits review as "Approved", commenting with areas that were tested.
+ 
+#### Stage 2: Dev Review
+
+Two developers will review the pull request and either give feedback or approve the PR. Any comments should be addressed before the pull request moves on to the next stage.
+
+ - PM reviewer adds `2: Dev Review` label, removes `1: PM Review` and `Setup Test Server` labels, and assigns 2 developers for review under Reviewers.
+   - At least one dev is assigned based on their [feature area](https://docs.mattermost.com/developer/core-developer-handbook.html#current-core-developers). Devs re-assign as needed.
+ - Devs review the code and provide feedback, with initial review completed within 2 business days. Some areas to check include:
+   - Proper Unit Tests
+   - API documentation
+   - Localization
+ - After the submitter has addressed and satisfied all reviewers' comments, `3: Ready to Merge` label is applied.
+
+#### Stage 3: Ready to Merge
+
+Review process is complete and the pull request is merged.
+
+ - Dev assigns `3: Ready to Merge` label.
+ - If Mattermost is not in release mode (between [major feature cut and release candidate cut](https://docs.mattermost.com/process/release-process.html)), the PR is merged into `master`.
+ - If the PR is a major change, merge is postponed until the next release cycle.
+   - Dev calls out on the issue that it is a major change and it will be merged after branching.
+   - Once the current release is branched the PR can be merged into `master`.
+
+#### PR Merged
+
+After a PR is merged:
+- External Contributions: PM closes the [Help Wanted] issue and related JIRA ticket.
+- Internal Contributions: Core committer resolves the JIRA ticket.
+- PM follows up for docs and changelog, and QA for release tests.
```

Comment:
@jasonblais What are the absolute minimum things a contributor would need to know _before_ submitting a pull request?

Does specific knowledge of the labels system and how they're used by the PMs qualify as required knowledge that would preclude a contributor from submitting a PR?

What about other steps of the review process? Do they _need_ to know all steps? Some steps? Any steps at all?

Reply:
I can think about it some more, but I just copied this file from https://github.com/mattermost/mattermost-server/blob/master/CONTRIBUTING.md

We do have plans to make the contribution guidelines brief, an early draft can be found here http://mattermost-developer-documentation.s3-website-us-east-1.amazonaws.com/commits/ca4b55e329e9e4c22fb71b41b45f748053795b76/contribute/getting-started/contribution-checklist/

Reply:
The review process could be separated into its own page and linked from CONTRIBUTING.md if a contributor is interested to learn more about it.

Reply:
Ah so this is just a copy/paste of the existing docs with a new process in the works: sounds like some good improvements are in the works and that's great! 👍